### PR TITLE
fix(reactivity): resolve Python scopes correctly in the dependency analyzer

### LINKED ISSUE #512: Reactive execution misses globals read inside functions, producing silently stale output
## Summary

A global variable read only inside a function body creates no dependency edge when the global is defined in a different block. Running the defining block with dependents therefore schedules nothing, and downstream blocks keep showing a successful but stale result.

Confirmed on Deepnote Cloud 2026-08-14.

## Reproduction

```python
# 1. Run all blocks. Then change old to new and run this block with dependencies.
source_value = "old"
```

```python
# 2. This function reads a global defined in block 1.
def read_source():
    return source_value
```

```python
# 3. This output should update after block 1 changes.
observed = read_source()
print(f"observed={observed!r}")
```

1. Run all three blocks with `source_value = "old"` — block 3 prints `observed='old'`.
2. Change block 1 to `"new"` and run block 1 with dependents.

## Actual / Expected

Actual: nothing downstream runs. Block 3 still prints `observed='old'` while the variable explorer shows `source_value = new`.

Expected: reactive execution follows free-variable reads inside callables and reruns their consumers.

The symptom is invisible on a full run, since every block re-executes regardless of edges. It also requires the value to actually change.

The dependency graph shows block 1 as an isolated node with a `source_value` output chip and no outgoing edge; block 2 has no input chips. The only edge is block 2 → block 3 via `read_source`.

## Root cause

Analyzer `packages/reactivity/src/scripts/ast-analyzer.py`, shipped as `@deepnote/reactivity`. `visit_Name` records a load only when the visitor is at global scope, or when the name is already in `self.global_vars`:

```python
if isinstance(node.ctx, ast.Load):
    if node.id in BUILTINS_SET:
        self.generic_visit(node)
        return
    if self.current_scope_is_global():
        self.used_global_vars.add(node.id)
    elif node.id in self.global_vars:
        self.used_global_vars.add(node.id)
```

`self.global_vars` is per-block. A visitor instance analyzes one block, so a global defined in a different block is never in that set. A free variable read inside a function body is therefore invisible precisely when its definition lives in another block — the only case a cross-block dependency graph exists to model.

Direct run of `VariableVisitor`:

```
b1 source_value        defined=['source_value'] used=[]
b2 def read_source     defined=['read_source'] used=[]      # source_value missing
b3 caller              defined=['observed'] used=['read_source']
same-block global      defined=['f','g'] used=['g']         # works within one block
method reads global    defined=['C'] used=[]                # same defect via a class method
```

Reads inside a function whose global lives in the same block still work, which is why this is easy to miss in testing.

## Provenance

Introduced in December 2024 when the analyzer started filtering function-local names out of used/defined variables; before that every load was tracked unconditionally:

```python
if isinstance(node.ctx, ast.Load):
    if node.id not in dir(builtins):
        self.used_vars.add(node.id)
```

A cosmetic fix (function locals appearing in `outputVariables`) introduced a correctness regression. For a dependency graph that is the wrong trade: an extra edge costs a redundant re-run, a missing edge costs correctness. A fix restoring unconditional load tracking was written and approved at the time but never merged, and the code has since migrated to the `@deepnote/reactivity` package carrying the regression with it.

## Impact

Downstream blocks display a successful but stale result while the variable explorer shows the new value. Reactive execution silently under-runs; the user has no signal that anything is out of date.

## Suggested fix

Track loads at any depth, and subtract names actually bound in the enclosing scope chain (function and lambda parameters, comprehension targets, local assignments).

## Related

See the companion issue on comprehension targets and lambda parameters being published as notebook variables: #513.

Both are the same underlying gap — the analyzer approximates Python scope with a per-block `global_vars` set, and block boundaries are not scope boundaries. This issue is the under-reporting direction (silently stale output); the companion issue is the over-reporting direction (phantom edges, spurious re-execution). A correct fix handles both directions at once: track loads at any depth, and subtract names actually bound in the enclosing scope chain. Fixing either direction alone risks re-introducing the other.

### LINKED ISSUE #513: Dependency graph publishes comprehension targets and lambda parameters as notebook variables
## Summary

The dependency-graph analyzer treats names bound inside a comprehension or a lambda as module-level notebook variables.

- A comprehension target is published as both an input and an output of its block. Downstream blocks reading a genuine module-level variable of the same name get their edge sourced from the comprehension block instead of the real definer.
- A lambda parameter is published as an input of its block, producing a false edge from whichever earlier block defines that name.

Kernel results stay correct; only the graph and the reactive re-execution it drives are wrong.

Confirmed on Deepnote Cloud 2026-08-14.

## Reproduction

```python
# 1. Real module-level value.
o = 5
```

```python
# 1b. Same name as the lambda parameter in the next block.
x = 99
```

```python
# 2. o and x are local bindings, not notebook variables.
items = [1, 2, 3]
squares = [o * o for o in items]
double = lambda x: x * 2
```

```python
# 3. Consumer: reads o, squares, double.
print(f"global o={o}; squares={squares}; double(3)={double(3)}")
```

1. Run all blocks.
2. Open the dependency graph.

## Actual / Expected

Actual:

- Block 2 lists `o` and `x` as inputs, with edges from block 1 and block 1b — although every `o` in block 2 is bound by the comprehension and `x` is bound as the lambda parameter. Neither global is read.
- Block 2 lists `o` as an output, alongside the genuine `double`, `items`, `squares`.
- Block 3's `o` edge is sourced from block 2, not block 1.

Expected: names bound by a comprehension or a lambda stay local to it and never appear in the block's input or output variables.

Runtime disagrees with the graph and stays correct: `global o=5; squares=[1, 4, 9]; double(3)=6`.

## Root cause

Analyzer `packages/reactivity/src/scripts/ast-analyzer.py`, shipped as `@deepnote/reactivity`, spawned as a `python3` child process; edges are built in `buildDagFromBlocks` (`src/dag-builder.ts`).

`VariableVisitor` tracks scope with a `scope_stack` pushed and popped in exactly three visitors: `visit_ClassDef`, `visit_FunctionDef`, `visit_AsyncFunctionDef`. There is no `visit_Lambda` and no comprehension visitor, so `Lambda`, `ListComp`, `SetComp`, `DictComp` and `GeneratorExp` bodies are walked with an empty scope stack and everything inside them is classified as module level.

- Comprehension targets are `ast.Name` with `ctx=Store`, so they hit the same branch as a real module-level assignment and land in `definedVariables` and `usedVariables`.
- Lambda parameters are `ast.arg` nodes, which never reach `visit_Name` — they escape `definedVariables` incidentally rather than by design — but their loads inside the body still land in `usedVariables`.

Direct run of `VariableVisitor`:

```
block2       defined=['double','items','o','squares']  used=['items','o','x']
def-scope    defined=['f']                             used=[]
lambda-only  defined=['double']                        used=['x']
comp-only    defined=['o','squares']                   used=['items','o']
```

## Impact

The graph is wrong and reactive mode re-runs blocks it should not: editing an unrelated same-named variable marks the comprehension block stale and cascades to everything downstream of it. This costs time on expensive blocks but never produces a wrong result.

A leaked name becomes a visible edge only when an earlier block defines the same name — and the leaking names are loop and lambda names such as `x`, `i`, `o`, `row`, `df`, which are exactly the ones reused across a notebook. Expect this to be under-reported rather than rare.

## Suggested fix

Push and pop a scope for `Lambda`, `ListComp`, `SetComp`, `DictComp` and `GeneratorExp` bodies, and treat comprehension targets and lambda parameters as bindings in that scope so they are subtracted from both the block's inputs and its outputs.

## Related

See the companion issue on globals read inside functions being missed, producing silently stale output: #512.

Both are the same underlying gap — the analyzer approximates Python scope with a per-block `global_vars` set, and block boundaries are not scope boundaries. This issue is the over-reporting direction (phantom edges, spurious re-execution); the companion issue is the under-reporting direction (silently stale output). A correct fix handles both directions at once: track loads at any depth, and subtract names actually bound in the enclosing scope chain (function and lambda parameters, comprehension targets, local assignments). Fixing either direction alone risks re-introducing the other.

### PR DESCRIPTION
**Pull request by jamesbhobbs**
## Problem

The dependency-graph analyzer (`packages/reactivity/src/scripts/ast-analyzer.py`) approximated Python scope with a per-block `global_vars` set and a stack of scope *names*. Block boundaries are not scope boundaries, so this went wrong in both directions:

- **Under-reporting (#512).** A global read inside a function or method body was only recorded if the global was defined in the *same* block. A global defined in another block produced no edge, so "run with dependents" scheduled nothing and downstream blocks kept showing a successful but stale result.
- **Over-reporting (#513).** There was no scope for `Lambda`, `ListComp`, `SetComp`, `DictComp` or `GeneratorExp`, so comprehension targets were published as block inputs *and* outputs, and lambda parameters as inputs. Same-named loop variables (`x`, `i`, `o`, `row`, `df`) then produced phantom edges and spurious re-execution.

The same root cause also dropped reads in decorators, default arguments and annotations, because those were walked inside the function's scope.

## Change

The visitor now tracks scope as a stack of *sets of locally bound names*, and a load is a module-level read whenever no enclosing scope binds it, at any depth:

- **Functions, async functions, lambdas** push a scope containing their parameters plus every name bound anywhere in the body (`_bound_names`, which does not descend into nested scopes). Pre-collecting is necessary because Python binds a name for the whole function: `x = x + 1` reads the local, not a global. Decorators, defaults, annotations and return annotations are visited in the enclosing scope before the push.
- **Comprehensions and generator expressions** visit the first iterable in the enclosing scope, then push a scope with their targets for the remaining iterables, conditions and element expressions.
- **Class bodies** push a scope for their own assignments; methods still cannot see class attributes, matching Python.
- `global` declarations continue to bypass the local check, so the existing `global num` behaviour is unchanged.

Analyzer diff is ~95 lines changed. No TypeScript source changes; `dag-builder.ts` consumes the same output shape.

## Tests

- `ast-analyzer.test.ts`: a `Python scoping` table of 15 cases covering both issues' repros plus the surrounding edge cases (method bodies, attribute reads, read-before-assign locals, `global`, decorators/defaults/annotations, class attributes, nested/dict/set/generator comprehensions, lambda defaults).
- `dag.test.ts`: two end-to-end DAG tests mirroring the issue reproductions — the missing `source_value` edge from #512 now exists, and the phantom `o`/`x` edges from #513 do not.
- Full suite: 3163 tests pass; `pnpm typecheck` and `pnpm biome:check` clean.

Not touched: `scripts/test-ast-analyzer.py` fails identically on `main` because its expected output predates the `importedPackages` and `linesOfCode` fields. It is not part of `pnpm test`; left for a separate cleanup.

Fixes #512
Fixes #513

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved Python dependency analysis across functions, classes, lambdas, comprehensions, decorators, annotations, and default arguments.
  - Correctly handles local bindings and `global` declarations when determining variable usage.
  - Improved dependency relationships for code blocks that reference variables defined elsewhere, including code nested within separate scopes.
  - Excludes comprehension targets and lambda parameters from unrelated dependency relationships.
  - Improved handling of class-level bindings so they do not incorrectly affect variable resolution inside methods.

### CODE REVIEW

File: packages/reactivity/src/scripts/ast-analyzer.py [hunk 1]
```diff
@@ -17,104 +17,129 @@ def __init__(self) -> None:
...
         return not self.scope_stack
 
-    def visit_Global(self, node):
-        for name in node.names:
-            self.function_globals.add(name)
-        self.generic_visit(node)
+    def _is_local(self, name):
+        if name in self.function_globals:
+            return False
+        return any(name in scope for scope in self.scope_stack)
```
[the remark this answers is not in the document]
Comment (jamesbhobbs) [reply packages/reactivity/src/scripts/ast-analyzer.py]: Fixed in 3719867. Scopes are now `(names, is_class)` tuples and `_is_local` skips class scopes unless innermost, so a method reading a name the class body also binds records the module-level read. Added two cases to the `Python scoping` table: a method reading a class-bound name, and a comprehension inside a class body. Please resolve if this addresses it.

File: packages/reactivity/src/scripts/ast-analyzer.py [hunk 2]
```diff
@@ -17,104 +17,129 @@ def __init__(self) -> None:
...
-    def visit_AnnAssign(self, node):
-        target = node.target
-        if isinstance(target, ast.Name) and self.current_scope_is_global():
-            self.global_vars.add(target.id)
-        self.generic_visit(node)
+    def _record_store(self, name):
+        if self.current_scope_is_global():
+            self.global_vars.add(name)
+        elif name not in self.function_globals:
+            self.scope_stack[-1].add(name)
```
[the remark this answers is not in the document]
Comment (jamesbhobbs) [reply packages/reactivity/src/scripts/ast-analyzer.py]: Done in 3719867. This was pre-existing behaviour rather than a regression, but it is the right call for the graph: `_record_store` now adds names declared `global` to `global_vars`, with a table case asserting `def f(): global g; g = 1` defines `g`. Please resolve if this addresses it.

**Comment by jamesbhobbs** [comment]
@coderabbitai the RUF005 nitpick is also applied in 3719867 (`[node.args, *body]`).

**Review by dinohamzic — CHANGES_REQUESTED** [review]
Ran a review against these changes + deepnote-internal and found this:

---

**Request changes.** I built this branch and ran `deepnote-internal`’s reactivity DAG tests against it: **30/30 passed**. The reactivity package tests also passed: **88/88**. The internal tests do **not** cover the changed Python scoping cases below, so these passes do not catch the regressions.

- **[P1] Comprehension walrus assignments lose an output** (`ast-analyzer.py:109`). `out = [(x := i) for i in items]` defines `x` at module scope, but the new analyzer reports only `out`. A later block reading `x` can be omitted from downstream execution.
- **[P2] A class-body read can lose its global dependency** (`ast-analyzer.py:74`). In `def f(): global x; class C: y = x; x = 2`, Python reads global `x` for `y`, but local HEAD reports no read. This finding applies to the local commit that is one commit ahead of the remote PR branch.
- **[P2] Defining a function falsely reports a global output** (`ast-analyzer.py:38`). `def f(): global x; x = 2` reports `x` as produced before `f` is called. That creates an incorrect DAG edge and extra reruns.

**Breaking-change assessment for current `deepnote-internal` usage:** The PR changes no exported TypeScript API or result shape, and `deepnote-internal` currently pins `@deepnote/reactivity@1.0.2`, so it has no immediate effect there. **If upgraded without these fixes, yes: the missing dependencies are a behavioral breaking change** that can leave notebook or published-app outputs stale. **With these fixes, I see no breaking change for the current internal call sites.** The DAG and rerun sets will still change by design to reflect Python scoping more accurately, but the adapter needs no API change.
